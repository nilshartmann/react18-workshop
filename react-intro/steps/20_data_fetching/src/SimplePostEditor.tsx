import React, { useEffect } from "react";
import TagChooser from "./TagChooser.tsx";

const ALL_TAGS = ["React", "TypeScript", "JavaScript"];

// 1. Render Phase
// 2. Commit Phase

let x = 0;

export default function SimplePostEditor() {
	const [title, setTitle] = React.useState("");
	const [body, setBody] = React.useState("");
	const [tags, setTags] = React.useState<string[]>([]);

	const clearDisabled = !title && !body && tags.length === 0;

	function clear() {
		setTitle("");
		setBody("");
		setTags([]);
	}

	function handleTitleChange(t: string) {
		setTitle(t);
	}

	return (
		<div>
			<h1>Create Post, aber simple!</h1>
			{/*<title>...</title>*/}

			<label>
				Title
				<input value={title} onChange={e => handleTitleChange(e.target.value)} />
			</label>
			{title.length === 0 && <p className={"error"}>Please enter a title</p>}

			<label>
				Body
				<textarea value={body} onChange={e => setBody(e.target.value)} />
			</label>
			{body.length === 0 && <p className={"error"}>Please enter a body</p>}

			<TagChooser availableTags={ALL_TAGS} selectedTags={tags} onSelectionChange={setTags} />

			<button disabled={clearDisabled} onClick={clear}>
				Clear
			</button>
			<button onClick={() => {}}>Save</button>
		</div>
	);
}
