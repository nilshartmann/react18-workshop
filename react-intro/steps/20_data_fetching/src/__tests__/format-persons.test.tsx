import { expect, test, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SimplePostEditor from "../SimplePostEditor";
const persons = [
  { firstname: "Klaus", age: 32 },
  { firstname: "Susi", age: 37 }
];

const greet = (name: string) => {
  return "Hello, " + name;
}

test("format persons", () => {
  const greeting = greet("Klaus")
  expect(greeting).toBe("Hello, Klaus");
});

test("renders correctly mit snapshot", () => {

  const { asFragment } = render(<SimplePostEditor/>);
  expect(asFragment()).toMatchSnapshot();

});


test("renders correctly", async () => {
  const user = userEvent.setup();
  render(<SimplePostEditor/>);

  screen.logTestingPlaygroundURL();

//   expect(screen.findByRole("button", {
//     name: /clear/i
// })).toBeInTheDocument();

  expect(screen.getByRole('button', { name: /clear/i })).toBeInTheDocument();
  expect(screen.getByRole('button', { name: /clear/i })).toBeDisabled();

  const titleInput = screen.getByLabelText("Title");
  const bodyInput = screen.getByLabelText("Body");
  await user.type(titleInput, "Hallo Welt");
  expect(titleInput).toHaveValue("Hallo Welt");

  expect(screen.getByRole('button', { name: /clear/i })).toBeEnabled();

  await user.click(screen.getByRole('button', { name: /clear/i }));
  expect(screen.getByRole('button', { name: /clear/i })).toBeDisabled();



});

