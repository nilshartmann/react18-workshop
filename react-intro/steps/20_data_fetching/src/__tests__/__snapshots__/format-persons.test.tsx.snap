// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Create Post
    </h1>
    <label>
      Title
      <input
        value=""
      />
    </label>
    <p
      class="error"
    >
      Please enter a title
    </p>
    <label>
      Body
      <textarea />
    </label>
    <p
      class="error"
    >
      Please enter a body
    </p>
    <div>
      <h2 />
      <div
        class="TagChooser__tags"
      >
        <label>
          <input
            type="checkbox"
          />
          React
        </label>
        <label>
          <input
            type="checkbox"
          />
          TypeScript
        </label>
        <label>
          <input
            type="checkbox"
          />
          JavaScript
        </label>
      </div>
    </div>
    <button
      disabled=""
    >
      Clear
    </button>
    <button>
      Save
    </button>
  </div>
</DocumentFragment>
`;

exports[`renders correctly mit snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Create Post, aber simple!
    </h1>
    <label>
      Title
      <input
        value=""
      />
    </label>
    <p
      class="error"
    >
      Please enter a title
    </p>
    <label>
      Body
      <textarea />
    </label>
    <p
      class="error"
    >
      Please enter a body
    </p>
    <div>
      <h2 />
      <div
        class="TagChooser__tags"
      >
        <label>
          <input
            type="checkbox"
          />
          React
        </label>
        <label>
          <input
            type="checkbox"
          />
          TypeScript
        </label>
        <label>
          <input
            type="checkbox"
          />
          JavaScript
        </label>
      </div>
    </div>
    <button
      disabled=""
    >
      Clear
    </button>
    <button>
      Save
    </button>
  </div>
</DocumentFragment>
`;
